### PR TITLE
OpenMetrics: force input to end with # EOF

### LINKED ISSUE #1169: Deal with incomplete scrapes.
A scrape might be incomplete because of a network error (connection closed, time-out, incomplete HTTP content, ...) or because a client wasn't able to collect all metrics (for whatever reason).

With https://github.com/prometheus/common/issues/14, it will be able to detect incomplete scrapes even without a network error.

At the very least, an incompletely scraped target must be marked as unhealthy, and its `up` metric must be 0. (This might or might not already be the case for network failures.)

Other measures to deal with incomplete scrapes need vetting:

It is easiest to reason about metrics if scrapes are atomic, i.e. either ingest all samples from a complete scrape, or none at all. Implementing an atomic scrape semantics has to be done carefully to not introduce a lot of allocation churn or unpredictable memory consumption (due to buffering until completion of a scrape is confirmed) and to not incur a significant performance cost.

On the other hand, in some situations it might be best to ingest what we can get, even if it is from an incomplete scrape. Via the `up` metric, alerts sensitive to incomplete scrapes can still be protected.

Note that a metric family will always be ingested as a whole from protobuf, even without a termination marker in the exchange format. The same guarantee can be given for the text format once we have a termination marker in the exchange format. However, since appending happens sequentially per time series, a rule evaluation might see some samples from a metric family already ingested but others not - even from a complete scrape.

### PR DESCRIPTION
**Pull request by roidelapluie**
Fixes #1169

Signed-off-by: Julien Pivotto <roidelapluie@inuits.eu>

### CODE REVIEW

File: pkg/textparse/openmetricsparse.go [hunk 1]
```diff
@@ -81,15 +81,17 @@ func (l *openMetricsLexer) Error(es string) {
...
+	series         []byte
+	text           []byte
+	mtype          MetricType
+	val            float64
+	ts             int64
+	hasTS          bool
+	start          int
+	offsets        []int
+	prev           token
+	ignoreEofError bool // only used to simplify tests.
```
Comment (brian-brazil) [inline pkg/textparse/openmetricsparse.go]: No need to do this, the tests already allow for this.

Comment (roidelapluie) [reply pkg/textparse/openmetricsparse.go]: Without it all the tests errors are "data does not end with # EOF"

Comment (roidelapluie) [reply pkg/textparse/openmetricsparse.go]: I have updated the pull request to use the public methods only

Comment (brian-brazil) [reply pkg/textparse/openmetricsparse.go]: Update the testcases then

Comment (roidelapluie) [reply pkg/textparse/openmetricsparse.go]: Updated

Comment (roidelapluie) [reply pkg/textparse/openmetricsparse.go]: Now it looks close to the way we test the Python parser

File: pkg/textparse/openmetricsparse_test.go [hunk 2]
```diff
@@ -272,71 +272,91 @@ func TestOpenMetricsParseErrors(t *testing.T) {
 		err   string
 	}{
 		{
-			input: "",
+			input: "# EOF",
```
Comment (brian-brazil) [inline pkg/textparse/openmetricsparse_test.go]: This is valid input.

File: pkg/textparse/openmetricsparse_test.go [hunk 3]
```diff
@@ -272,71 +272,91 @@ func TestOpenMetricsParseErrors(t *testing.T) {
...
 			err:   "EOF",
 		},
+		{
+			input: "# EOF\n",
+			err:   "EOF",
```
Comment (brian-brazil) [inline pkg/textparse/openmetricsparse_test.go]: This is also valid input.

File: pkg/textparse/openmetricsparse_test.go (hunk 2 above)
Comment (roidelapluie) [reply pkg/textparse/openmetricsparse_test.go]: Yes, the error is EOF; which means that it is valid. It was like it before.

File: pkg/textparse/openmetricsparse_test.go (hunk 3 above)
Comment (roidelapluie) [reply pkg/textparse/openmetricsparse_test.go]: Same err. We test the error code of the next() and in this case it is EOF.

**Comment by roidelapluie** [comment]
My understanding is that we tested the err returned by Next(). In that situation EOF is returned by Next() ; I did not change the way tests are written.

File: pkg/textparse/openmetricsparse.go [hunk 4]
```diff
@@ -232,13 +233,18 @@ func (p *OpenMetricsParser) Next() (Entry, error) {
...
+	t := p.nextToken()
+	defer func() { p.prev = t }()
+	switch t {
 	case tEofWord:
 		if t := p.nextToken(); t != tEOF {
 			return EntryInvalid, errors.New("unexpected data after # EOF")
 		}
 		return EntryInvalid, io.EOF
 	case tEOF:
+		if p.prev != tEofWord {
```
Comment (bwplotka) [inline pkg/textparse/openmetricsparse.go]: nit: We tend to capitalize all acronyms on Go e.g `EOF` as described [here](https://github.com/golang/go/wiki/CodeReviewComments#initialisms)

File: pkg/textparse/openmetricsparse_test.go (hunk 2 above)
Comment (brian-brazil) [reply pkg/textparse/openmetricsparse_test.go]: This needs a comment then.

File: pkg/textparse/openmetricsparse.go (hunk 4 above)
Comment (roidelapluie) [reply pkg/textparse/openmetricsparse.go]: Except that this variable is generated by Golex -- and the link you provide says:

> Code generated by the protocol buffer compiler is exempt from this rule. Human-written code is held to a higher standard than machine-written code.

While this is not protocol buffer compiler, it is machine generated code.

File: pkg/textparse/openmetricsparse_test.go (hunk 2 above)
Comment (roidelapluie) [reply pkg/textparse/openmetricsparse_test.go]: Added

**Comment by brian-brazil** [comment]
Thanks!

**Pull request merged**